src/drivers/imu/mpu9250/MPU9250_mag.cpp

```diff
@@ -200,92 +124,15 @@ MPU9250_mag::_measure(struct ak8963_regs data)
 		perf_count(_mag_overflows);
 	}
 
-	mag_report	mrb;
-	mrb.timestamp = hrt_absolute_time();
-
-	mrb.is_external = _parent->is_external();
+	_px4_mag.set_external(_parent->is_external());
+	_px4_mag.set_temperature(_parent->_last_temperature);
+	_px4_mag.set_error_count(perf_event_count(_mag_errors));
 
 	/*
 	 * Align axes - note the accel & gryo are also re-aligned so this
 	 *              doesn't look obvious with the datasheet
 	 */
-	float xraw_f, yraw_f, zraw_f;
-
-	mrb.x_raw =  data.x;
-	mrb.y_raw = -data.y;
-	mrb.z_raw = -data.z;
-
-	xraw_f =  data.x;
-	yraw_f = -data.y;
-	zraw_f = -data.z;
-
-	/* apply user specified rotation */
-	rotate_3f(_parent->_rotation, xraw_f, yraw_f, zraw_f);
-
-	mrb.x = ((xraw_f * _mag_range_scale * _mag_asa_x) - _mag_scale.x_offset) * _mag_scale.x_scale;
-	mrb.y = ((yraw_f * _mag_range_scale * _mag_asa_y) - _mag_scale.y_offset) * _mag_scale.y_scale;
-	mrb.z = ((zraw_f * _mag_range_scale * _mag_asa_z) - _mag_scale.z_offset) * _mag_scale.z_scale;
-	mrb.scaling = _mag_range_scale;
-	mrb.temperature = _parent->_last_temperature;
-	mrb.device_id = _parent->_mag->_device_id.devid;
-
-	mrb.error_count = perf_event_count(_mag_errors);
-
-	_mag_reports->force(&mrb);
-
-	/* notify anyone waiting for data */
-	if (mag_notify) {
-		poll_notify(POLLIN);
-	}
-
-	if (mag_notify && !(_pub_blocked)) {
-		/* publish it */
-		orb_publish(ORB_ID(sensor_mag), _mag_topic, &mrb);
-	}
-}
-
-int
-MPU9250_mag::ioctl(struct file *filp, int cmd, unsigned long arg)
-{
-	/*
-	 * Repeated in MPU9250_accel::ioctl
-	 * Both accel and mag CDev could be unused in case of magnetometer only mode or MPU6500
-	 */
-
-	switch (cmd) {
-
-	case SENSORIOCRESET:
-		return ak8963_reset();
-
-	case SENSORIOCSPOLLRATE: {
-			switch (arg) {
-
-			/* zero would be bad */
-			case 0:
-				return -EINVAL;
-
-			case SENSOR_POLLRATE_DEFAULT:
-				return ioctl(filp, SENSORIOCSPOLLRATE, MPU9250_ACCEL_DEFAULT_RATE);
-
-			/* adjust to a legal polling interval in Hz */
-			default:
-				return _parent->_set_pollrate(arg);
-			}
-		}
-
-	case MAGIOCSSCALE:
-		/* copy scale in */
-		memcpy(&_mag_scale, (struct mag_scale *) arg, sizeof(_mag_scale));
-		return OK;
-
-	case MAGIOCGSCALE:
-		/* copy scale out */
-		memcpy((struct mag_scale *) arg, &_mag_scale, sizeof(_mag_scale));
-		return OK;
-
-	default:
-		return (int)CDev::ioctl(filp, cmd, arg);
-	}
+	_px4_mag.update(timestamp_sample, data.x * _mag_asa_x, -data.y * _mag_asa_y, -data.z * _mag_asa_z);
```

Comment:
Hi @dagar , I think the multiplications in this line should be accomplished prior to the method call rather than in the arguments. This helps adhere to a general philosophy of one operation per instruction. This is outlined in Clean Code chapter 3, a section called "Do One Thing" as well as Bjarne Stroustrup's sentiment in the JSF coding standards..  If you want bugs, hide them like this.  I think that operations should be separate from arguments, always. :)

Reply:
Thanks @mcsauder. It looks like the additional sensitivity scale factors per axis are relatively common for mags (hmc5883 separation calibration, etc). I'll push it into PX4Magnetometer.